waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -637,20 +512,6 @@ proc queryLoop(w: WakuStore, query: HistoryQuery, candidateList: seq[RemotePeerI
     debug "failed to resolve the query"
     return err("failed to resolve the query")
 
-proc findLastSeen*(list: seq[IndexedWakuMessage]): float = 
-  var lastSeenTime = float64(0)
-  for iwmsg in list.items : 
-    if iwmsg.msg.timestamp>lastSeenTime: 
-      lastSeenTime = iwmsg.msg.timestamp 
-  return lastSeenTime
-
-proc isDuplicate(message: WakuMessage, list: seq[WakuMessage]): bool =
```

Comment:
This is the proc I mentioned in https://github.com/status-im/nim-waku/pull/849#discussion_r807421594

Reply:
Right. The new `contains` check will still consider both message `payload` and `contentTopic` (since both are hashed into the digest). It also considers the `timestamp` (via `senderTimestamp`).

Reply:
> I'd suggest using pubsubTopic + contentTopic + message digest + sender timestamp as a composite Primary key in the DB.

Ah, thanks, this is very helpful in terms of data model design! Afaict the `contentTopic` is already included in the message digest (see above), so it would be `pubsubTopic` + `digest` + `senderTimestamp`. If this is the combination that guarantees absolute uniqueness, this should also be the combination that forms the `Index` - the `Index` functions analagous to the SQLite primary key in the in-memory storage: it has to be sufficient to guarantee uniqueness (but, of course, also _minimal_)

> I do not recommend falling back to the receiver timestamp, it counts as an integrity violation.

Right, and I agree with your suggestions if we follow this approach, thanks! _But_ this requires a change in specification (at least for `store`), where the `timestamp` is currently optional in https://rfc.vac.dev/spec/14/ and `receiverTime` is specified as an optional fallback in https://rfc.vac.dev/spec/13/. And I think this is a fairly significant change in specification, because we will now require every Waku v2 node to be run on a system where it has access to a synchronised clock - at least if it wants its messages to be stored. Is this a tradeoff we can make?
If we mandate use of `senderTime` for store purposes, I don't think it needs to affect `relay` (`relay` should forward messages regardless of whether `timestamp` is populated or not).

Reply:
> this should also be the combination that forms the Index - the Index functions analagous to the SQLite primary key in the in-memory storage: it has to be sufficient to guarantee uniqueness (but, of course, also minimal)

The `Index` structure has no trace of pubsub topic since afaik there is only one active pubsub topic i.e., the default waku pubsub, so adding that field to Index would not be useful and just redundant info to be saved. That is why in the initial implementation, the pubsub topic was not part of the Index. However, I agree that it makes sense to have it so that the Index reflects the DB primary key composition.

Reply:
>Right, and I agree with your suggestions if we follow this approach, thanks! This requires a change in specification (at least for store), though, where the timestamp is currently optional in https://rfc.vac.dev/spec/14/ and receiverTime is specified as an optional fallback in https://rfc.vac.dev/spec/13/.

The receiver ts can be used for sorting but the spec does not suggest using it to make up for the lack of sender ts.

>  And I think this is a fairly significant change in specification because we will now require every Waku v2 node to be run on a system where it has access to a synchronized clock - at least if it wants its messages to be stored. Is this a tradeoff we can make?

We do have a similar assumption in the FT-store protocol for the sake of time range queries.  Also, clocks do not have to be perfectly synchronized, a margin of 20 seconds is considered. I am not sure how big of a trade-off it is. 

Also, mandating timestamps for waku messages is essential for their availability in the store protocol. 
Consider a store node that sorts messages using sender ts, and receives messages with no timestamp, then such messages will be placed somewhere at the beginning of the message history (since they point to the past) hence with a high probability won't be served to the requesters especially for the time range queries (like in FT-store protocol). So, such messages will be almost effectively discarded and not served properly even if store nodes do not want to discard them.

Reply:
We may skip mandating waku message timestamp and just can explain what would be the consequence of the lack of it.

Reply:
Posted in PM, but can we please separate out performance from spec changes?

If a spec decision or similar is required to merge PR, we should label this PR as `blocked` with link to spec discussion. The spec issue should summarize problem, context, proposed solution(s) and consequences.

If this PR is blocked by a wider spec discussion I don't think it makes sense to block this PR even further https://github.com/vacp2p/rfc/pull/483#issuecomment-1042403189